File: src/ltea/EPCgw.cpp
/*******************************************************************************
 * This file is part of IMTAphy
 * _____________________________________________________________________________
 *
 * Copyright (C) 2011
 * Institute of Communication Networks (LKN)
 * Department of Electrical Engineering and Information Technology (EE & IT)
 * Technische Universitaet Muenchen
 * Arcisstr. 21
 * 80333 Muenchen - Germany
 * http://www.lkn.ei.tum.de/~jan/imtaphy/index.html
 * 
 * _____________________________________________________________________________
 *
 *   IMTAphy is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   IMTAphy is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with IMTAphy.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

// based on code from openWNS with the following license:

/*******************************************************************************
 * This file is part of openWNS (open Wireless Network Simulator)
 * _____________________________________________________________________________
 *
 * Copyright (C) 2004-2007
 * Chair of Communication Networks (ComNets)
 * Kopernikusstr. 5, D-52074 Aachen, Germany
 * phone: ++49-241-80-27910,
 * fax: ++49-241-80-22242
 * email: info@openwns.org
 * www: http://www.openwns.org
 * _____________________________________________________________________________
 *
 * openWNS is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Lesser General Public License version 2 as published by the
 * Free Software Foundation;
 *
 * openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <IMTAPHY/ltea/EPCgw.hpp>
#include <IMTAPHY/ltea/pdcp/eNB.hpp>

#include <DLL/Layer2.hpp>
#include <DLL/StationManager.hpp>

#include <WNS/module/Base.hpp>

using namespace ltea;

STATIC_FACTORY_REGISTER_WITH_CREATOR(EPCgw,
                                     wns::node::component::Interface,
                                     "ltea.EPCgw",
                                     wns::node::component::ConfigCreator);

EPCgw::EPCgw(wns::node::Interface* node, const wns::pyconfig::View& _config) :
    dll::RANG(node, _config),
    config(_config),
    logger(config.get("logger"))
{
    MESSAGE_SINGLE(NORMAL, logger, "LTEA::EPCgw created");
    
    // to allow filtering for UL/DL in probes by identfying the probing node: 0==EPC, 1==UE, 2==BS
    node->getContextProviderCollection().addProvider(wns::probe::bus::contextprovider::Constant("NodeType", 0));
}

void
EPCgw::sendData(
           const wns::service::dll::UnicastAddress& _peer,
           const wns::osi::PDUPtr& pdu,
           wns::service::dll::protocolNumber protocol,
           wns::service::dll::FlowID _epsBearer)
{
    MESSAGE_SINGLE(NORMAL, logger, "Outgoing PDU with EPS Radio Bearer: "<<_epsBearer<<" for UE: "<<_peer);
    
    if (accessPointLookup.knows(_peer))
    {
        dll::RANG::sendData(_peer, pdu, protocol, _epsBearer);


    }
    else
    {
        MESSAGE_SINGLE(NORMAL, logger, "No eNB registered that knows UE " << _peer);
        MESSAGE_SINGLE(NORMAL, logger, "Dropping PDU.");
    }
}

void 
EPCgw::setAssociatedUEs(std::vector< wns::node::Interface* > associatedUEs, ltea::pdcp::PDCPeNB* pdcp)
{
    for (std::vector<wns::node::Interface*>::const_iterator iter = associatedUEs.begin();
         iter != associatedUEs.end(); iter++)
    {        
        dll::UpperConvergence* dataTransmissionService = 
            (*iter)->getService<dll::UpperConvergence*>(config.get<std::string>("ueDllDataTransmissionServiceName"));
        
        assure(dataTransmissionService, "Could not get remote DLL data transmission service from UE");    
            
        wns::service::dll::UnicastAddress macAddress = dataTransmissionService->getMACAddress();
        
        
        MESSAGE_SINGLE(NORMAL, logger, "Registering UE "<< (*iter)->getName() << " at " << pdcp->getFUN()->getLayer()->getNodeName());
        accessPointLookup.insert(macAddress, pdcp);

    }
}

File: modules/phy/imtaphy/PyConfig/ltea/nodes/EPCgw.py
###############################################################################
# This file is part of openWNS (open Wireless Network Simulator)
# _____________________________________________________________________________
#
# Copyright (C) 2004-2007
# Chair of Communication Networks (ComNets)
# Kopernikusstr. 16, D-52074 Aachen, Germany
# phone: ++49-241-80-27910,
# fax: ++49-241-80-22242
# email: info@openwns.org
# www: http://www.openwns.org
# _____________________________________________________________________________
#
# openWNS is free software; you can redistribute it and/or modify it under the
# terms of the GNU Lesser General Public License version 2 as published by the
# Free Software Foundation;
#
# openWNS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
###############################################################################

import openwns.node
import dll.Layer2

import constanze.node
import tcp.TCP
import ltea.nodes.IPv4Component
import ip.AddressResolver

class Tunnel( dll.Layer2.Layer2 ):
    dataTransmission = None
    notification = None
    flowIDUnbindDelay = 60.0 # was 60.0
    dllDataTransmissions = None
    dllNotifications = None
    ueDllDataTransmissionServiceName = None

    def __init__(self, node, parentLogger = None):
        dll.Layer2.Layer2.__init__(self, node, "EPCgw", parentLogger = parentLogger)
        self.dataTransmission = "EPCgw.dllDataTransmission"
        self.notification = "EPCgw.dllNotification"
        self.nameInComponentFactory = "ltea.EPCgw"
        self.dllDataTransmissions = []
        self.dllNotifications = []
        self.logger.enabled=True


    def setUEdllDataTransmissionServiceName(self, serviceName):
        self.ueDllDataTransmissionServiceName = serviceName

    def addBS(self, bs):
        self.dllDataTransmissions.append(openwns.node.FQSN(bs, bs.dll.dataTransmission))
        self.dllNotifications.append(openwns.node.FQSN(bs, bs.dll.notification))

class EPCgw(openwns.node.Node):
    tunnel = None
    nl  = None
    load = None
    tl = None

    def __init__(self, probeWindow, settlingTime, useTCP=True):
        super(EPCgw, self).__init__("EPCgw")

        self.setProperty("Type", "EPCgw")

        # create Tunnel to reach the BSs
        self.tunnel = Tunnel(self, self.logger)
        address = 256*255 - 1
        self.tunnel.setStationID(address)

        # create Network Layer
        domainName = "EPCgw.ltea.wns.org"
        
        self.nl = ltea.nodes.IPv4Component.IPv4Component(self, domainName + ".ip",domainName, probeWindow, settlingTime, useDllFlowIDRule = True)
        self.nl.addDLL(_name = "tun",
                       # Where to get my IP Address
                       _addressResolver = ip.AddressResolver.FixedAddressResolver("192.168.254.254", "255.255.0.0"),
                       # ARP zone
                       _arpZone = "LTEARAN",
                       # We can deliver locally
                       _pointToPoint = False,
                       # DLL service names
                       _dllDataTransmission = self.tunnel.dataTransmission,
                       _dllNotification = self.tunnel.notification)

        if useTCP == True:
            self.tl = tcp.TCP.TCPComponent(self, "tcp", self.nl.dataTransmission, self.nl.notification)
        else:
            self.tl = tcp.TCP.UDPComponent(self, "udp", self.nl.dataTransmission, self.nl.notification)

        # create Loadgen
        self.load = constanze.node.ConstanzeComponent(self, domainName + ".constanze", self.logger)


    def addTraffic(self, binding, load):
        self.load.addTraffic(binding, load)
